Step2 - 경로 조회 기능

mins99 opened the pull request:
안녕하세요 리뷰어님!
step1의 피드백을 마저 반영하고 step2 구현하여 리뷰 요청 드립니다.

이전 ATDD 미션을 진행했을 때는 인수 테스트에 집중하느라 서비스 로직에 대한 테스트를 신경쓰지 못했는데
이번 미션에서 추가해보았습니다! 첫 mockito 사용이라 어려웠네요 ㅎㅎ;
처음 TDD 미션을 하기전과 비교하면 도메인, 서비스, 인수테스트까지 다양한 테스트로 정말 튼튼한 안전장치를 만든 느낌입니다!

그리고 주어진 미션 수행 순서에 최대한 맞춰 진행해보았습니다(Outside In)
다만 이부분은 어짜피 수정이 될텐데 하면서 커밋을 미루다보니 실제 커밋 순서는 파일 생성 순서의 반대로 되어버렸네요
다음 미션에서는 커밋 순서도 조금 신경써서 해보겠습니다 😅

예외 상황에 대한 테스트의 경우 어디에 구현해야 할지 고민했는데 우선 서비스 테스트에 모두 구현했습니다.
최대한 다양한 테스트 클래스에 추가하는 것이 좋을까요? 아니면 적절한 위치가 있을까요?

이번 미션도 잘부탁드립니다 🥰

mins99 committed:
refactor(hasAllStation): equals 비교로 수정

- 상행역과 하행역의 동일 유무는 equals 로 처리 가능하다

mins99 committed:
docs(README): 2단계 - 경로 조회 기능 요구 사항 정리

mins99 committed:
feat(findShortestPath): 최단 경로 조회 인수 테스트

- Happy 케이스에 대한 구현

mins99 committed:
fix(Section): 중간 구간 추가시 거리값 조정이 안되는 오류 수정

mins99 committed:
feat(Line): 구간 조회 로직 추가

mins99 committed:
refactor(Response): 응답 객체 불변으로 변경

- 추가로 응답 포맷 맞추는 작업도

mins99 committed:
feat(PathRequest): 최단 경로 조회 요청 객체 구현

mins99 committed:
feat(PathResponse): 최단 경로 조회 응답 객체 구현

mins99 committed:
feat(PathFinder): 최단 경로 조회 도메인 구현

- jgrapht 라이브러리를 활용

mins99 committed:
feat(PathService): 최단 경로 조회 기능 구현

mins99 committed:
test(PathServiceTest): 최단 경로 조회 기능 테스트 추가

- 성공 케이스와 실패 케이스 추가

mins99 committed:
feat(PathController): 최단 경로 조회 기능 구현

- 예외 처리 추가

mins99 committed:
test(PathAcceptanceTest): 최단 경로 조회 인수 테스트

- setup 로직 일부 수정

mins99 committed:
docs(README): 2단계 - 경로 조회 기능 구현 사항 체크

mins99 committed:
docs(README): 2단계 - 경로 조회 기능 구현 사항 체크

mins99 commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -0,0 +1,52 @@
+
+    public GraphPath getShortestPath(final Station sourceStation, final Station targetStation) {
+        validInputCheck(sourceStation, targetStation);
+        GraphPath shortestPath = dijkstraShortestPath.getPath(sourceStation, targetStation);
```
예외 사항 3가지 중 `존재하지 않는 출발역이나 도착역을 조회 할 경우` 와 관련하여
DijkstraShortestPath 클래스의 getPath 메서드에 자체적으로 존재하지 않는 역에 대한 예외 처리가 구현되어 있었습니다
그래서 따로 구현을 안했는데 예외 처리가 되어 있어도 추가로 구현을 하는게 나을까요?
```java
if (!graph.containsVertex(source)) {
      throw new IllegalArgumentException("graph must contain the source vertex");
}
if (!graph.containsVertex(sink)) {
      throw new IllegalArgumentException("graph must contain the sink vertex");
}
```

dhmin5693 replied:
예외 처리는 그대로 써도 좋긴 합니다.
별도의 validation을 수행한다면 오류상황에서 확인하기 더 쉬운 메시지를 보내줄 수 있을거에요.

dhmin5693 commented on src/main/java/nextstep/subway/line/dto/LineResponse.java:
```diff
@@ -18,7 +17,8 @@ public class LineResponse {
     }
 
-    private LineResponse(Long id, String name, String color, List<StationResponse> stations, LocalDateTime createdDate, LocalDateTime modifiedDate) {
+    private LineResponse(final Long id, final String name, final String color, final List<StationResponse> stations,
```
내부 값 설정에만 사용하고 있어서 매개변수에 final이 없어도 되지 않을까요?

그리고 아래는 리뷰가 아니라 궁금한 점인데요,
이번 미션에 부쩍 final 매개변수를 많이 사용하셨더라구요.
혹시 특별한 이유가 있으실까요?

final 매개변수의 장점은 매개변수에 새 주소나 값을 할당하지 않도록 강제하는 건데 저는 큰 장점이 없다고 생각하거든요.

mins99 replied:
음.. 좀 더 명확하게 불변 데이터임을 표시하고자 final을 붙였습니다
지나친 final은 오히려 독이려나요 ㅎㅎ;
final 사용과 관련해서 조금 더 고민해보겠습니다!

dhmin5693 commented on src/main/java/nextstep/subway/path/ui/PathController.java:
```diff
@@ -0,0 +1,42 @@
+
+    @GetMapping
+    public ResponseEntity<PathResponse> findShortestPath(@RequestParam Long source, @RequestParam Long target) {
+        PathRequest pathRequest = new PathRequest(source, target);
```
이 프로젝트에서 `*request`는 보통 외부에서 데이터가 인입될 때 처음으로 사용되고 있어요.
생성해서 서비스로 넘기기보다는 애초에 이 메소드의 매개변수로 받는건 어떤가요?

mins99 replied:
`@RequestBody`를 사용할 때만 가능한줄 알았는데, 어노테이션 없이 객체로 선언해도 바로 저장이 되네요?!
수정하도록 하겠습니다!

dhmin5693 commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -0,0 +1,52 @@
+
+public class PathFinder {
+
+    private final WeightedMultigraph<Station, DefaultWeightedEdge> graph = new WeightedMultigraph(DefaultWeightedEdge.class);
```
`WeightedMultigraph`, `DijkstraShortestPath` 등 제네릭이 적용 가능한 클래스엔 타입을 꼭 기입해주세요.
잘못된 데이터를 혼용하지 않을 수 있습니다.

mins99 replied:
`제네릭이 적용 가능한 클래스엔 타입을 꼭 기입`
해당 부분도 수정하겠습니다! 처음 써보는 클래스라 익숙하지 않았네요 ㅎㅎ;

dhmin5693 commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -0,0 +1,52 @@
+        dijkstraShortestPath = new DijkstraShortestPath<>(graph);
+    }
+
+    public GraphPath getShortestPath(final Station sourceStation, final Station targetStation) {
```
`PathFinder`는 2단계 미션에서 가장 중요한 클래스에요.
TDD 미션이니만큼 확실하게 검증할 수 있는 테스트를 작성해주세요.

mins99 replied:
넵 관련 테스트 추가하겠습니다

dhmin5693 commented on src/test/java/nextstep/subway/path/application/PathServiceTest.java:
```diff
@@ -0,0 +1,135 @@
+    @Test
+    @DisplayName("조회하려는 최단 구간의 출발역과 도착역은 달라야 한다")
+    void 출발역과_도착역이_같은_경우() {
+        // when
+        when(stationService.findStationById(1L)).thenReturn(교대역);
+        when(lineRepository.findAll()).thenReturn(Arrays.asList(이호선, 삼호선, 신분당선));
```
84번째 줄의 주석이 // when 으로 기록이 되어 있는데요,
보통은 given - when - then라고 합니다.

given에선 테스트의 조건이나 환경을 설정하고 when에서 수행한 결과를 then에서 검증하는 것입니다.

**엥? 근데 왜 메소드는 when이죠?** 라는 의문을 품으실 수 있어요.
그래서 `org.mockito.BDDMockito.given`을 많이들 쓰고 있습니다.
덜 헷갈려서 좋거든요.

```suggestion
        // given
        given(stationService.findStationById(1L)).willReturn(교대역);
        given(lineRepository.findAll()).willReturn(Arrays.asList(이호선, 삼호선, 신분당선));
```

mins99 replied:
각 객체의 결과가 해당 값일때 라는 의미로 when에 넣었었는데 그 보단 주어지는 값이라는 의미로 given에 넣는것이 맞는거군요 😲`org.mockito.BDDMockito.given` 을 적용해서 수정해보겠습니다!

dhmin5693 commented on src/test/java/nextstep/subway/path/application/PathServiceTest.java:
```diff
@@ -0,0 +1,135 @@
+     */
+    @BeforeEach
+    public void setUp() {
+        pathService = new PathService(lineRepository, stationService);
```
여기에서 생성해주는 대신 34라인(테스트 대상)의 위에 `@InjectMocks`를 기입하는 방법도 있으니 참고해주세요 :)

mins99 replied:
바로 적용해보겠습니다!

`@Mock`과 `@InjectMocks`의 차이도 찾아보았습니다
-> UserController에 대한 단위 테스트를 작성하고자 할 때, UserService를 사용하고 있다면 
`@Mock` 어노테이션을 통해 가짜 UserService를 만들고, `@InjectMocks`를 통해 UserController에 이를 주입시킬 수 있다.
[참고](https://mangkyu.tistory.com/145)

dhmin5693 replied:
저번에 코멘트로 남겼던 test double 중 fake 기법을 활용할 수 있는 `@Spy`도 있어요.
`@InjectMocks` 에 같이 주입됩니다.

dhmin5693 commented:
이번 미션에서 핵심이 되는 인수 테스트는 통합 테스트의 일환으로, 특정한 request를 사용하여 API를 호출하면 기대하는 결과가 반환되는지는 확인하는 것입니다.
특정 로직이 아니라 하나의 액션에 소모되는 모든 인프라까지 통합하여 전부 검증하는 셈이지요.

반면 이번에 공부하셨다는 mockito 기반의 서비스 테스트는 단위테스트에요.
외부의 결과를 고정시키고 해당 로직을 수행하면서 나타날 수 있는 모든 케이스를 검증하는게 목표입니다.

단위테스트에서 왜 mock으로 외부의 결과를 고정시킬까요?
과학시간에 배운 실험군과 대조군을 생각해보시면 이해가 쉽습니다. ([링크](https://ko.wikipedia.org/wiki/실험군과_대조군))
단위 테스트는 실험군입니다.
대조군의 결과가 고정되어 있을 때 반드시 수행해야 하는 무언가가 있고 그걸 검증하는게 단위 테스트인거죠.
우리는 대조군의 결과를 조금씩 바꾸어가며(다른 데이터를 mocking) 실험군이 어떠한 결과를 도출하는지를 테스트로 검증할 수도 있겠죠?\

여기서 대조군을 지정하는 방법은 mock만 있는건 아니에요.
fake 객체를 만들어서 별도의 클래스로 특정 결과를 지정해버릴 수도 있습니다.
이러한 내용을 통틀어서 `test double`이라고 합니다.
아래 URL에 잘 기재되어 있으니 확인해보시면 좋겠네요 ㅎㅎ
https://tecoble.techcourse.co.kr/post/2020-09-19-what-is-test-double/


이번에 테스트를 작성하면서 예외를 서비스단에서 테스트를 하셨는데요, 서비스에서 일어날 수 있는 예외이기 때문에 테스트 대상이 맞습니다.
위치가 매우 적절하니 이 점에 대한 걱정은 안하셔도 좋을 것 같아요 ㅎㅎ

추가로 그 예외가 실제로 일어나는 부분 또한 테스트를 작성해주시면 더 좋습니다.
조금 과하다 생각하기 때문에 리뷰에도 남기지 않고 추천드리지는 않지만, 극단적인 연습을 해보는 방법의 일환으로 **테스트 커버리지 100% 달성** 같은 목표를 세우고 실행에 옮겨볼 수도 있어요.

dhmin5693 reviewed (changes requested):
두 번째 미션 수행하시느라 굉장히 수고 많으셨습니다.

로직을 구현하는게 쉽지는 않은 미션인데 굉장히 잘 구현해주셨네요.
로직 외적 요소에 작은 피드백 위주로 남겨두었으니 확인 부탁드릴게요.

mins99 committed:
refactor(findShortestPath): PathRequest 객체를 매개변수로 받아 전달하도록 수정

mins99 committed:
refactor(PathFinder): 제네릭이 적용 가능한 클래스 제네릭 타입 명시

mins99 committed:
test(PathServiceTest): InjectMocks 적용

mins99 committed:
test(PathFinderTest): PathFinder 객체 테스트 추가

mins99 committed:
test(PathServiceTest): when 메서드를 given 으로 변경

- 테스트의 조건이나 환경을 설정한다는 점에서 when 메서드는 given 의 의미로 사용되었음
- 헷갈리지 않도록 BDDMockito의 given 으로 변경

mins99 committed:
chore(LineRequest): 미사용 메서드 삭제

mins99 committed:
refactor(Response): 응답 객체 불변 키워드 삭제

- 내부 값 설정에만 사용하는 객체이므로 final 키워드가 필요하지 않음

mins99 commented:
안녕하세요 리뷰어님!
코멘트 주신 부분들 반영 완료하여 리뷰 요청드립니다 😊
미션이 진행될 수록 신경 쓸 부분이 더 많아지지만 그만큼 많이 알아가는 느낌입니다 🥳 (많이 부족함도 느낍니다 흑흑)

`PathFinderTest` 와 관련하여 PathService의 findPath를 검증하는 부분과 PathFinder의 getShortestPath를 검증하는 부분이 동일한 로직을 검증한다는 생각으로 따로 구현을 안했었는데 말씀주셔서 추가했습니다! 
생각해보니 각 클래스별로 테스트를 만들지 않는다면 리팩토링이나 요구사항 변경으로 인해 로직이 수정될 때 테스트의 보호를 받을 수 없겠네요 😢
(도메인 TDD는 정말 빙산의 일각이었구나 느낍니다.. test double 도 좀 더 학습해보겠습니다!)

리뷰어님의 꼼꼼한 리뷰 덕분에 많이 공부하고 있습니다 넘 감사드려요 🥰 
수정 사항들 확인 부탁드릴게요~!

dhmin5693 reviewed (approved):
피드백 반영 잘해주셨어요 😄

중요한 로직은 소중한 보물이라고 생각하며 테스트로 잘 보호해주세요.
이번 단계는 마무리하고 다음 미션 진행해주세요.
